### PR TITLE
Add `pymatgen.io.pwmat` module

### PR DESCRIPTION
**Pull request by lhycms**
## Summary

PWmat is a GPU-accelerated first-principles calculation software developed under the leadership of Professor Lin-wang Wang. I am currently utilizing it for my research project, and in the course of this work, I have written some scripts. I hope to submit them to pymatgen to assist more people in conveniently handling data in the PWmat format. This submission includes only structure file and trajectory file, and I plan to continue developing code for other input and output files in subsequent submissions.

Major changes:

- feature 1: Add `pymatgen.io.pwmat.inputs.AtomConfig` which is input/output structure file, work as POCAR in vasp.
- feature 2: Add `pymatgen.io.pwmat.outputs.Movement` which is trajectory file generated when running MD.
- feature 3: Add pwmat fmt within Structure.from_file() and Structure.from_str().
- feature 4: Add object for output files of PWmat: OUT.FERMI, REPORT, DOS.totalspin and so on.

## Todos

In the future, I will contribute code processing other input and output files for PWmat.

### CODE REVIEW

**Comment by janosh** [comment]
Thanks a lot! This is much appreciated! 👍 

We also need extensive tests for these three new features.

File: tests/files/pwmat/DOS.spinup_projected
Comment (janosh) [inline tests/files/pwmat/DOS.spinup_projected:1]: Can we compress all test files with 1k+-lines?

File: pymatgen/io/pwmat/outputs.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        chunk_sizes_bak.insert(0, 0)
+        chunk_starts: list[int] = np.cumsum(chunk_sizes_bak).tolist()
+        chunk_starts.pop(-1)
+        return chunk_sizes, chunk_starts
+
+    @property
+    def atom_configs(self):
+        """
+        Returns:
+            atom_configs: List[Structure]
```
Comment (janosh) [inline pymatgen/io/pwmat/outputs.py:67]: return type should come first, then description:

        Returns:
           list[Structure]: atom_configs

File: pymatgen/io/pwmat/outputs.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        """
+        return [step["atom_config"] for _, step in enumerate(self.ionic_steps)]
+
+    @property
+    def etots(self):
```
Comment (janosh) [inline pymatgen/io/pwmat/outputs.py:72]: please add return type annotations to all functions and methods:

def etots(self) -> np.ndarray

**Review by janosh — CHANGES_REQUESTED** [review]

File: pymatgen/io/pwmat/outputs.py (hunk 2 above)
Comment (lhycms) [reply pymatgen/io/pwmat/outputs.py:72]: Thank you for pointing out the issue. I will revise it as soon as possible.

**Comment by lhycms** [comment]
Hi, jonash. I have made the modifications according to your feedback and further added comments. It is my first PR in github, I'm not familiar with this😂

Please check if there are any unreasonable things, i will continue to modify it.

File: pymatgen/core/structure.py [hunk 3]
```diff
@@ -2746,6 +2746,10 @@ def to(self, filename: str | Path = "", fmt: FileFormats = "", **kwargs) -> str:
                 with zopen(filename, "wt", encoding="utf8") as file:
                     file.write(res_str)
             return res_str
+        elif fmt == "pwmat" or fnmatch(filename.lower(), "*.pwmat") or fnmatch(filename.lower(), "*.config"):
```
Comment (janosh) [inline pymatgen/core/structure.py:2749]: can you add a test for this where we do a round trip with `Structure.from_file` and back with `Structure.to(file)`?

**Review by janosh — CHANGES_REQUESTED** [review]
I renamed a bunch of method names and dict keys to use snake_case. Let me know if you're ok with that. there's one more test missing for the new `Structure` `pwmat` IO format that needs a test. othewise this is ready to go.

File: pymatgen/core/structure.py (hunk 3 above)
Comment (lhycms) [reply pymatgen/core/structure.py:2749]: When I use 'python -m unittest discover -s <pymatgen_dir>/tests/io/pwmat' for testing, the program will report the error shown in the following figure. How can I correctly test the 'pymatgen.pwmat.io' module?

截屏2024-01-17 下午8 53 35

Comment (janosh) [reply pymatgen/core/structure.py:2749]: We use `pytest`, so do `pip install pytest && pytest tests/io/pwmat`.

Comment (lhycms) [reply pymatgen/core/structure.py:2749]: Thanks for help!

**Review by janosh — APPROVED** [review]
This looks great now. Thanks a lot @lhycms for this impressive contribution! 👍

**Comment by lhycms** [comment]
Thank you for your guidance and assistance. You are really nice😄. @janosh

**Pull request merged**